src/Cardano/Wallet/CoinSelection/Fee.hs

```diff
@@ -341,3 +352,128 @@ computeFee (CoinSelection inps outs chgs) =
     -- by definition, impossible; unless we messed up real hard.
     collapse [] _ =
         invariant "outputs are bigger than inputs" (undefined) (const False)
+
+-- | A linear equation on the transaction size. Represents the @\s -> a + b*s@
+-- function where @s@ is the transaction size in bytes, @a@ and @b@ are
+-- constant coefficients.
+data TxSizeLinear =
+    TxSizeLinear (Quantity "lovelace" Double) (Quantity "lovelace/byte" Double)
+    deriving (Eq, Show)
+
+data AddressScheme = Sequential | Random
+    deriving (Eq, Show)
+
+data Network = Mainnet | Testnet
+    deriving (Eq, Show)
+
+cardanoPolicy :: TxSizeLinear
+cardanoPolicy = TxSizeLinear (Quantity 155381) (Quantity 43.946)
+
+-- | Estimate the fee for a transaction that has @inps@ inputs
+-- and @outps@ outputs as coming from coin selection.
+--
+-- NOTE: The average size of @Attributes AddrAttributes@ and
+--       the transaction attributes @Attributes ()@ are both hard-coded
+estimateCardanoFee
+    :: TxSizeLinear
+    -> AddressScheme
+    -> Network
+    -> CoinSelection
+    -> Fee
+estimateCardanoFee (TxSizeLinear (Quantity a) (Quantity b)) scheme net (CoinSelection inps outs chngs) =
+    Fee $ ceiling (a + b*totalPayload)
+    where
+
+        --input     ---------------------------------- 42-43
+        --  | list len 2                  -- 1
+        --  | word8                       -- 1
+        --  | tag 24                      -- 2
+        --  | bytes ------------------------ 2 + 36
+        --  |   | list len 2  -- 1
+        --  |   | bytes       -- 2 + 32
+        --  |   | word32      -- 1-2
+        sizeOfTxIn :: Int
+        sizeOfTxIn = 42
+
+        -- For sequential scheme
+        -- output    ---------------------------------- 48-56 (mainnet) & 56-63 (testnet)
+        --  | list len 2                  -- 1
+        --  | sizeOf(address)             -- 46-54
+        --  | word64                      -- 1-8
+
+        -- address   ---------------------------------- 46 (mainnet) & 54 (testnet)
+        --  | list len 2                  -- 1
+        --  | tag 24                      -- 2
+        --  | bytes          --------------- 2 + 37-45
+        --  |   | list len 3 -- 1
+        --  |   | bytes      -- 2 + 32
+        --  |   | attributes -- 1-8
+        --  |   | word8      -- 1
+        --  | word32                      -- 1-4
+
+        -- For random scheme
+        -- output    ---------------------------------- 69-77 (mainnet) & 77-85 (testnet)
+        --  | list len 2                  -- 1
+        --  | sizeOf(address)             -- 67-75
+        --  | word64                      -- 1-8
+
+        -- address   ---------------------------------- 67 (mainnet) & 75 (testnet)
+        --  | list len 2                  -- 1
+        --  | tag 24                      -- 2
+        --  | bytes          --------------- 2 + 48-56
+        --  |   | list len 3 -- 1
+        --  |   | bytes      -- 2 + 43
+        --  |   | attributes -- 1-8
```

Comment:
I believe comment is wrong here, that's not the size of the root hash that is bigger for the random scheme, but the size of the attributes.

Reply:
corrected